katdal/caldata.py

```diff
@@ -0,0 +1,400 @@
+                                                                            x)
+                    y_im[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                            self._y[:, ch_idx, pol_idx, ant_idx].imag,
+                                                                            x)
+        y = y_re + y_im
+        return y
+
+
+class ComplexInterpolate1D(object):
+    def __init__(self, x, y):
+        self._x = x
+        self._y = y
+        self._mag = np.abs(y)
+        self._phase = y / self._mag
+
+    def __call__(self, x):
+        mag = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        phase_re = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        phase_im = np.empty((x.shape[0], self._y.shape[1], self._y.shape[2], self._y.shape[3]))
+        # Interpolate across nans in time axis.
+        for ch_idx in range(self._y.shape[1]):
+            for pol_idx in range(self._y.shape[2]):
+                for ant_idx in range(self._y.shape[3]):
+                    mag[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                           self._mag[:, ch_idx, pol_idx, ant_idx],
+                                                                           x)
+                    phase_re[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                                self._phase[:, ch_idx, pol_idx, ant_idx].real,
+                                                                                x)
+                    phase_im[:, ch_idx, pol_idx, ant_idx] = _safe_linear_interp(self._x,
+                                                                                self._phase[:, ch_idx, pol_idx, ant_idx].imag,
+                                                                                x)
+        phase = phase_re + phase_im
+        return phase / np.abs(phase) * mag
+
+
+def interpolate_nans_1d(y):
+    """Interpolate over nans in a timeseries.
+
+    When interpolating the per-channel timeseries, the channels with more
+    nan values will deviate from the neighboring channels appearing as spikes
+    when the mean over time is plotted.
+    """
+    if np.isnan(y).all():
+        return y  # nothing to do, all nan values
+    if np.isfinite(y).all():
+        return y  # nothing to do, all values valid
+
+    # interpolate across nans
+    nan_locs = np.isnan(y)
+    xi = np.nonzero(~nan_locs)[0]
+    yi = y[xi]
+    x = np.arange(len(y))
+    y = _safe_linear_interp(xi, yi, x)
+    return y
+
+
+def _get_cal_sensor(key, katdal_obj):
+    """Load a calibrator solution from sensors.
+    """
+    values = []
+    timestamps = []
+    for sensor in katdal_obj.file['TelescopeState'].keys():
+        if key in sensor:
```

Comment:
Can you explain why you're doing a substring match, instead of just looking up `key` directly into the sensor cache? There is no guarantee that some sensor name won't be a substring of another, unrelated sensor.